envi/archs/h8/emu.py

```diff
@@ -0,0 +1,1314 @@
+"""
+The H8 Emulator module.
+"""
+
+import sys
+import struct
+
+import emu
+import envi
+import envi.bits as e_bits
+from envi.archs.h8 import H8Module
+from envi.archs.h8.regs import *
+from disasm import H8RegDirOper
+
+# calling conventions
+class H8ArchitectureProcedureCall(envi.CallingConvention):
+    """
+    Implement calling conventions for your arch.
+    """
+    def execCallReturn(self, emu, value, ccinfo=None):
+        sp = emu.getRegister(REG_SP)
+        pc = struct.unpack(">H", emu.readMemory(sp, 2))[0]
+        sp += 2 # For the saved pc
+        sp += (2 * argc) # Cleanup saved args
+
+        emu.setRegister(REG_SP, sp)
+        emu.setRegister(REG_R0, value)
+        emu.setProgramCounter(pc)
+
+    def getCallArgs(self, emu, count):
+        return emu.getRegisters(0xf)  # r0-r3 are used to hand in parameters.  additional ph8s are stored and pointed to by r0
+
+aapcs = H8ArchitectureProcedureCall()
+
+
+CPUSTATE_RESET =    0
+CPUSTATE_EXC =      1
+CPUSTATE_EXEC =     2
+CPUSTATE_BUS =      3
+CPUSTATE_SLEEP =    4
+CPUSTATE_SWSTDBY =  5
+CPUSTATE_HWSTDBY =  6
+
+class H8Emulator(H8Module, H8RegisterContext, envi.Emulator):
+    IVT_RESET = 0
+
+    def __init__(self, advanced=True):
+        H8Module.__init__(self)
+        envi.Emulator.__init__(self, self)
+        H8RegisterContext.__init__(self)
+
+        self.state = CPUSTATE_RESET
+        self.ptrsz = 0
+
+        seglist = [ (0,0xffffffff) for x in xrange(6) ]
+
+        self.setAdvanced(advanced)
+        self.addCallingConvention("H8 Arch Procedure Call", aapcs)
+
+    def setAdvanced(self, advanced=True):
+        self.advanced = advanced
+        if advanced:
+            self.ptrsz = 4
+        else:
+            self.ptrsz = 2
+
+    def reset(self):
```

Comment:
This brings up ( again ) the need for 'interrupt' emulation in the emu state machine... :)  Probably best left to mk2? ( obvs, this special purpose method can stay, just discussing... )